src/com/phonebookservice/util/MyArrayList.java

```diff
@@ -92,7 +99,16 @@ public boolean addAll(final Collection<? extends T> collection) {
     @Override
     public boolean addAll(final int index,
             final Collection<? extends T> collection) {
-        throw new UnsupportedOperationException();
+        this.checkCollectionNotNull(collection);
+        final Iterator<? extends T> iterator = collection.iterator();
+        int startIndex = index;
+
+        while (iterator.hasNext()) {
+            this.add(startIndex, (T) iterator.next());
```

Comment:
How many shift are you doing here ??

If index = 0, and array has size million.

How many time you will visit those million element ??

Reply:
shift to all indices after a certain index.
will need to visit a million element in for each element